// ikagaka.balloon.js/BalloonSurface.js
// Generated by CoffeeScript 1.7.1
var BalloonSurface;

BalloonSurface = (function() {
  var SurfaceUtil;

  SurfaceUtil = window["SurfaceUtil"] || window["Ikagaka"]["SurfaceUtil"];

  function BalloonSurface(element, scopeId, balloonConf, balloons) {
    this.element = element;
    this.scopeId = scopeId;
    this.balloons = balloons;
    this.descript = balloonConf.descript;
    this.baseCanvas = balloonConf.canvas;
    this.render();
  }

  BalloonSurface.prototype.destructor = function() {
    $(this.element).off();
    return void 0;
  };

  BalloonSurface.prototype.render = function() {
    var type, util;
    type = this.scopeId === 0 ? "sakura" : "kero";
    util = new SurfaceUtil(this.element);
    util.init(this.baseCanvas);
    return void 0;
  };

  return BalloonSurface;

})();

if ((typeof module !== "undefined" && module !== null ? module.exports : void 0) != null) {
  module.exports = BalloonSurface;
}

if (window["Ikagaka"] != null) {
  window["Ikagaka"]["BalloonSurface"] = BalloonSurface;
}
