// uupaa.wmevent.js/lib/WMEvent.js
(function(global) {
"use strict";

// --- dependency modules ----------------------------------
// --- define / local variables ----------------------------
//var _runOnNode = "process" in global;
//var _runOnWorker = "WorkerLocation" in global;
//var _runOnBrowser = "document" in global;

// --- class / interfaces ----------------------------------
function WMEvent() {
    this._on    = {}; // event database. { type: [callback, ...], ... }
    this._types = []; // registered types. ["type", ...]
}

//{@dev
WMEvent["repository"] = "https://github.com/uupaa/WMEvent.js";
//}@dev

WMEvent["prototype"] = {
    "constructor":  WMEvent,            // new WMEvent()
    "register":     WMEvent_register,   // WMEvent#register(types:EventTypeStringArray):this
    "list":         WMEvent_list,       // WMEvent#list(type):FunctionArray
    "has":          WMEvent_has,        // WMEvent#has(type):Boolean
    "on":           WMEvent_on,         // WMEvent#on(target:Any|null, type:EventTypeString, callback:Function):this
    "off":          WMEvent_off,        // WMEvent#off(target:Any|null, type:EventTypeString, callback:Function):this
    "fire":         WMEvent_fire,       // WMEvent#fire(target:Any|null, type:EventTypeString):this
    "clear":        WMEvent_clear       // WMEvent#clear(target:Any|null):this
};

// --- implements ------------------------------------------
function WMEvent_register(types) { // @arg EventTypeStringArray - ["type", ...]
                                   // @ret this
                                   // @desc register types
//{@dev
    $valid($type(types, "Array"), WMEvent_register, "types");
//}@dev

//{@dev
    var that = this;

    // avoid duplicate
    types.forEach(function(type) {
        if (that._types.indexOf(type) < 0) {
            that._types.push(type);
        }
    });
//}@dev
    return this;
}

function WMEvent_list(type) { // @arg EventTypeString
                              // @ret FunctionArray - [callback, ...]
                              // @desc get registered callback functions array
//{@dev
    $valid($type(type, "String"), WMEvent_list, "type");
//}@dev

    if ( this["has"](type) ) {
        return this._on[type];
    }
    return [];
}

function WMEvent_has(type) { // @arg EventTypeString
                             // @ret Boolean
                             // @desc has registered type
//{@dev
    $valid($type(type, "String"), WMEvent_has, "type");
//}@dev

    return type in this._on;
}

function WMEvent_on(target,     // @arg Any|null - event target object
                    type,       // @arg EventTypeString
                    callback) { // @arg Function - callback(event:EventObject):void
                                // @ret this
                                // @desc add event listener
//{@dev
    $valid($type(type,     "String"),   WMEvent_on, "type");
    $valid($type(callback, "Function"), WMEvent_on, "callback");

    if (this._types.indexOf(type) < 0) {
        throw new TypeError("WMEvent#on(type) is not registered");
    }
//}@dev

    if ( this["has"](type) ) {
        this._on[type].push(callback);
    } else {
        this._on[type] = [callback];
    }
    if (target) {
        target["addEventListener"](type, callback, false);
    }
    return this;
}

function WMEvent_off(target,     // @arg Any|null - event target object
                     type,       // @arg EventTypeString
                     callback) { // @arg Function - callback(event:EventObject):void
                                 // @ret this
                                 // @desc remove event listener
//{@dev
    $valid($type(type,     "String"),   WMEvent_off, "type");
    $valid($type(callback, "Function"), WMEvent_off, "callback");
//}@dev

    if ( this["has"](type) ) {
        var pos = this._on[type].indexOf(callback); // [fn1, fn2, fn3, ...]

        if (pos >= 0) {
            this._on[type].splice(pos, 1);

            if (!this._on[type].length) {
                delete this._on[type];
            }
        }
    }
    if (target) {
        target["removeEventListener"](type, callback, false);
    }
    return this;
}

function WMEvent_fire(target, // @arg Any|null - event target object
                      type) { // @arg EventTypeString
                              // @ret this
                              // @desc fire event
//{@dev
    $valid($type(type, "String"), WMEvent_fire, "type");
//}@dev

    if (target && target["dispatchEvent"]) {
        target["dispatchEvent"]( _createDOMLevel4EventObject(type) );
    } else if ( this["has"](type) ) {
        this._on[type].forEach(function(callback) {
            callback( _createDOMLevel4EventObject(type) );
        });
    }
}

function WMEvent_clear(target) { // @arg Any|null
                                 // @ret this
    for (var type in this._on) {
        var callbacks = this._on[type];

        for (var i = 0, iz = callbacks.length; i < iz; ++i) {
            if (target) {
                target["removeEventListener"](type, callbacks[i], false);
            }
        }
    }
    this._on = {};
    return this;
}

function _createDOMLevel4EventObject(type) {
    if (global["Event"]) {
        return new global["Event"](type);
    }
    // http://www.w3.org/TR/dom/#interface-event

    // omitted properties of JScript.
    return {
            "type":             type,
            "target":           null,
            "currentTarget":    null,
            "bubbles":          false,
            "cancelable":       false,
            "defaultPrevented": false,
            "eventPhase":       0,
          //"cancelBubble":     false,  [JScript]
          //"clipboardData":    void 0, [JScript]
          //"returnValue":      true,   [JScript]
          //"srcElement":       null,   [JScript]
            "timeStamp":        Date.now()
        };
}

// --- validate / assertions -------------------------------
//{@dev
function $valid(val, fn, hint) { if (global["Valid"]) { global["Valid"](val, fn, hint); } }
function $type(obj, type) { return global["Valid"] ? global["Valid"].type(obj, type) : true; }
//function $keys(obj, str) { return global["Valid"] ? global["Valid"].keys(obj, str) : true; }
//function $some(val, str, ignore) { return global["Valid"] ? global["Valid"].some(val, str, ignore) : true; }
//function $args(fn, args) { if (global["Valid"]) { global["Valid"].args(fn, args); } }
//}@dev

// --- exports ---------------------------------------------
if ("process" in global) {
    module["exports"] = WMEvent;
}
global["WMEvent" in global ? "WMEvent_" : "WMEvent"] = WMEvent; // switch module. http://git.io/Minify

})((this || 0).self || global); // WebModule idiom. http://git.io/WebModule
